From 8b459916d8575cd78d29a413532b38d9563a2085 Mon Sep 17 00:00:00 2001
From: zheng qian <xqq@xqq.im>
Date: Sun, 25 Apr 2021 11:19:11 +0900
Subject: [PATCH v4 2/4] avformat/mpegts: Fix indentation for stream_id
 constants

Signed-off-by: zheng qian <xqq@xqq.im>
---
 libavformat/mpegts.h | 24 ++++++++++++------------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/libavformat/mpegts.h b/libavformat/mpegts.h
index 224282b3ed..c75ba08f29 100644
--- a/libavformat/mpegts.h
+++ b/libavformat/mpegts.h
@@ -138,18 +138,18 @@
 #define STREAM_TYPE_AUDIO_EAC3      0x87
 
 /* ISO/IEC 13818-1 Table 2-22 */
-#define STREAM_ID_PROGRAM_STREAM_MAP 0xbc
-#define STREAM_ID_PRIVATE_STREAM_1   0xbd
-#define STREAM_ID_PADDING_STREAM     0xbe
-#define STREAM_ID_PRIVATE_STREAM_2   0xbf
-#define STREAM_ID_AUDIO_STREAM_0     0xc0
-#define STREAM_ID_VIDEO_STREAM_0     0xe0
-#define STREAM_ID_ECM_STREAM         0xf0
-#define STREAM_ID_EMM_STREAM         0xf1
-#define STREAM_ID_DSMCC_STREAM       0xf2
-#define STREAM_ID_TYPE_E_STREAM      0xf8
-#define STREAM_ID_METADATA_STREAM    0xfc
-#define STREAM_ID_EXTENDED_STREAM_ID 0xfd
+#define STREAM_ID_PROGRAM_STREAM_MAP        0xbc
+#define STREAM_ID_PRIVATE_STREAM_1          0xbd
+#define STREAM_ID_PADDING_STREAM            0xbe
+#define STREAM_ID_PRIVATE_STREAM_2          0xbf
+#define STREAM_ID_AUDIO_STREAM_0            0xc0
+#define STREAM_ID_VIDEO_STREAM_0            0xe0
+#define STREAM_ID_ECM_STREAM                0xf0
+#define STREAM_ID_EMM_STREAM                0xf1
+#define STREAM_ID_DSMCC_STREAM              0xf2
+#define STREAM_ID_TYPE_E_STREAM             0xf8
+#define STREAM_ID_METADATA_STREAM           0xfc
+#define STREAM_ID_EXTENDED_STREAM_ID        0xfd
 #define STREAM_ID_PROGRAM_STREAM_DIRECTORY  0xff
 
 /* ISO/IEC 13818-1 Table 2-45 */
-- 
2.29.2

